finagle/buoyant/src/main/scala/com/twitter/finagle/buoyant/ExistentialStability.scala

```diff
@@ -20,33 +20,46 @@ object ExistentialStability {
      * will update.
      */
     def stabilizeExistence: Var[Option[Var[T]]] = {
-      val init = unstable.sample().map(Var(_))
-      Var.async[Option[VarUp[T]]](init) { update =>
-        // the current inner Var, null if the outer Var is None
-        @volatile var current: VarUp[T] = null
-        @volatile var exists = false
-        val mu = new {}
+
+      // The inner Var is null until the first time unstable becomes a Some.
+      @volatile var inner: Var[T] = null
+
+      def makeInner(init: T) = Var.async(init) { update =>
         unstable.changes.respond {
-          case Some(t) if current == null => mu.synchronized {
-            // T created
-            exists = true
-            current = Var(t)
-            update() = Some(current)
-          }
-          case Some(t) if !exists => mu.synchronized {
-            // T re-created
-            exists = true
-            current() = t
-            update() = Some(current)
-          }
-          case Some(t) => mu.synchronized {
-            // T modified
-            current() = t
-          }
-          case None => mu.synchronized {
-            // T deleted
-            exists = false
-            update() = None
+          case Some(t) => update() = t
+          case _ =>
+        }
+      }
+
+      // Initial value for outer Var.
+      val init = unstable.sample().map { t =>
+        if (inner == null) {
```

Comment:
TIOLI: It looks like we do this null check every time we call `makeInner`...maybe we should factor that out so that it's part of `makeInner`?

Reply:
gonna LI because `makeInner` returns a Var.  Moving the null check into `makeInner` means we have to move the assignment into `makeInner` as well which seems less clear to me.